File: src/main/java/cn/earthsky/rayattributes/attribute/AttributeType.java
package cn.earthsky.rayattributes.attribute;

public enum AttributeType {
    ATTACK("攻击力", false),
    ATTACK_PERCENT("攻击力%", true),
    HEALTH("生命", false),
    HEALTH_PERCENT("生命%", true),
    DEFENSE("防御", false),
    DEFENSE_PERCENT("防御%", true),
    CRIT_RATE("暴击率", true),
    CRIT_DAMAGE("暴击伤害", true),
    SKILL_DAMAGE("技能增伤", true),
    ENERGY_REGEN("能量恢复效率", true),
    COOLDOWN_REDUCTION("冷却缩减", true),
    DAMAGE_REDUCTION("伤害减免", true);

    private final String displayName;
    private final boolean percentage;

    AttributeType(String displayName, boolean percentage) {
        this.displayName = displayName;
        this.percentage = percentage;
    }

    public String getDisplayName() {
        return displayName;
    }

    public boolean isPercentage() {
        return percentage;
    }

    public String formatValue(double value) {
        if (percentage) {
            return String.format("%.1f%%", value * 100);
        }
        return String.valueOf((int) value);
    }
}

File: tools/attribute-sim/server/src/engine/AttributeType.ts
/** 属性类型 — 与插件 AttributeType.java 对齐 */
export enum AttributeType {
  ATTACK = 'ATTACK',
  ATTACK_PERCENT = 'ATTACK_PERCENT',
  HEALTH = 'HEALTH',
  HEALTH_PERCENT = 'HEALTH_PERCENT',
  DEFENSE = 'DEFENSE',
  DEFENSE_PERCENT = 'DEFENSE_PERCENT',
  CRIT_RATE = 'CRIT_RATE',
  CRIT_DAMAGE = 'CRIT_DAMAGE',
  SKILL_DAMAGE = 'SKILL_DAMAGE',
  ENERGY_REGEN = 'ENERGY_REGEN',
  COOLDOWN_REDUCTION = 'COOLDOWN_REDUCTION',
  DAMAGE_REDUCTION = 'DAMAGE_REDUCTION',
}

export const ATTR_META: Record<AttributeType, { displayName: string; isPercent: boolean }> = {
  [AttributeType.ATTACK]: { displayName: '攻击力', isPercent: false },
  [AttributeType.ATTACK_PERCENT]: { displayName: '攻击力%', isPercent: true },
  [AttributeType.HEALTH]: { displayName: '生命', isPercent: false },
  [AttributeType.HEALTH_PERCENT]: { displayName: '生命%', isPercent: true },
  [AttributeType.DEFENSE]: { displayName: '防御', isPercent: false },
  [AttributeType.DEFENSE_PERCENT]: { displayName: '防御%', isPercent: true },
  [AttributeType.CRIT_RATE]: { displayName: '暴击率', isPercent: true },
  [AttributeType.CRIT_DAMAGE]: { displayName: '暴击伤害', isPercent: true },
  [AttributeType.SKILL_DAMAGE]: { displayName: '技能增伤', isPercent: true },
  [AttributeType.ENERGY_REGEN]: { displayName: '能量恢复效率', isPercent: true },
  [AttributeType.COOLDOWN_REDUCTION]: { displayName: '冷却缩减', isPercent: true },
  [AttributeType.DAMAGE_REDUCTION]: { displayName: '伤害减免', isPercent: true },
};

export function fmtAttr(type: AttributeType, value: number): string {
  const meta = ATTR_META[type];
  if (meta.isPercent) return (value * 100).toFixed(1) + '%';
  return value.toFixed(0);
}

export function fmtAttrShort(type: AttributeType, value: number): string {
  const meta = ATTR_META[type];
  if (meta.isPercent) return (value * 100).toFixed(1) + '%';
  return value.toFixed(1);
}
